arrow/src/main/java/org/apache/iceberg/arrow/vectorized/VectorizedReaderBuilder.java

```diff
@@ -101,33 +99,26 @@ public VectorizedReader<?> message(
         Lists.newArrayListWithExpectedSize(icebergFields.size());
 
     for (Types.NestedField field : icebergFields) {
-      int id = field.fieldId();
-      VectorizedReader<?> reader = readersById.get(id);
-      if (idToConstant.containsKey(id)) {
-        reorderedFields.add(constantReader(field, idToConstant.get(id)));
-      } else if (id == MetadataColumns.ROW_POSITION.fieldId()) {
-        if (setArrowValidityVector) {
-          reorderedFields.add(VectorizedArrowReader.positionsWithSetArrowValidityVector());
-        } else {
-          reorderedFields.add(VectorizedArrowReader.positions());
-        }
-      } else if (id == MetadataColumns.IS_DELETED.fieldId()) {
-        reorderedFields.add(new DeletedVectorReader());
-      } else if (reader != null) {
-        reorderedFields.add(reader);
-      } else if (field.initialDefault() != null) {
-        reorderedFields.add(
-            constantReader(field, convert.apply(field.type(), field.initialDefault())));
-      } else if (field.isOptional()) {
-        reorderedFields.add(VectorizedArrowReader.nulls());
-      } else {
-        throw new IllegalArgumentException(
-            String.format("Missing required field: %s", field.name()));
-      }
+      VectorizedReader<?> reader =
+          VectorizedArrowReader.replaceWithMetadataReader(
+              field, readersById.get(field.fieldId()), idToConstant, setArrowValidityVector);
+      reorderedFields.add(defaultReader(field, reader));
```

Comment:
Same as the refactoring done in https://github.com/apache/iceberg/pull/12836/files